### PR TITLE
Add a capability and transport version for new regex and range interval query support

### PR DESCRIPTION
**Pull request by ChrisHegarty**
This commit adds a capability and transport version for new regex and range interval query support.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/query/IntervalsSourceProvider.java [hunk 1]
```diff
@@ -823,6 +823,14 @@ public String getWriteableName() {
 
         @Override
         public void writeTo(StreamOutput out) throws IOException {
+            if (out.getTransportVersion().before(TransportVersions.REGEX_AND_RANGE_INTERVAL_QUERIES)) {
+                throw new UnsupportedOperationException(
```
Comment (ChrisHegarty) [inline server/src/main/java/org/elasticsearch/index/query/IntervalsSourceProvider.java:827]: this seems a little redundant with the gating of this feature behind a capability - assuming that the capability ensures that all nodes in the cluster have it ?

Comment (mayya-sharipova) [reply server/src/main/java/org/elasticsearch/index/query/IntervalsSourceProvider.java:827]: As  I understood, a cluster can have nodes with different capabilities. But I don't understand is how does this gating even work? I think it is still possible for a node to send this query request to another node  that doesn't have this capability?  

Also, what is the benefit of throwing this error now, as opposed on a shard which will happen automatically where a shard discovers an unknown NamedWriteableRegistry.Entry?

**Review by mayya-sharipova — APPROVED** [review]
@ChrisHegarty Thanks for following this up and addressing. The change LGTM.

File: server/src/main/java/org/elasticsearch/index/query/IntervalsSourceProvider.java (hunk 1 above)
Comment (thecoop) [reply server/src/main/java/org/elasticsearch/index/query/IntervalsSourceProvider.java:827]: This check should be handled automatically by implementing `VersionedNamedWriteable`, and implementing the `getMinimalSupportedVersion` method

File: server/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java [hunk 2]
```diff
@@ -95,6 +95,11 @@ public List<Route> routes() {
         );
     }
 
+    @Override
+    public Set<String> supportedCapabilities() {
+        return SearchCapabilities.CAPABILITIES;
```
Comment (ChrisHegarty) [inline server/src/main/java/org/elasticsearch/rest/action/search/RestSearchAction.java:100]: Hmm... I think async_search is going to need to "inherit" the same set of capabilities - or at least this one.

**Comment by ChrisHegarty** [comment]
@elasticsearchmachine update branch

**Review by thecoop — APPROVED** [review]

**Pull request merged**